### PR TITLE
[Swift] Add closures concept exercise

### PR DESCRIPTION
**Pull request by wneumann**
Addresses issue https://github.com/exercism/v3/issues/1671.

### CODE REVIEW

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+That takes an integer, `base` and returns an optional closure. The function should return a closure that works just like `generateDigit` or `generateBit`, but for base `base` rather than just base 2 or 10.
+
+```swift
+let gen7 = generatorForBase(7)
+// => (Int) -> (seed: Int, next: Int)
+gen7!(123)
+// => (seed 17, next 4)
+```
+
+## 6. Generate a Collatz sequence from a positive integer
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:85]: The `collatz-conjecture` is an existing v2 exercise (see https://github.com/exercism/swift/blob/master/exercises/collatz-conjecture/README.md) and thus ineligible for usage in a Concept Exercise.

File: languages/swift/exercises/concept/closures/.meta/config.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "authors": [
+    {
+      "github_username": "wneumann",
+      "exercism_username": "wneumann"
+    }
+  ],
+  "contributors": [],
+  "forked_from": []
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.meta/config.json:9]: If the fields don't have any contents, you can omit them:

```suggestion
```

File: languages/swift/exercises/concept/closures/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+Closures in Swift are self-contained blocks of code that can be passed parameters to trigger their computation and return values. Closures also capture values from their environment and use them in their computations. As they are self contained, they may be passed around in a program like other values or assigned to constants and variables.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/introduction.md:1]:
```suggestion
Closures in Swift are self-contained blocks of code that can be passed parameters to trigger their computation and return values. Closures also capture values from their environment and can use them in their computations. As they are self contained, they may be passed around in a program like other values or assigned to constants and variables.
```

File: languages/swift/exercises/concept/closures/.docs/introduction.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+This may sound very familiar. Closures may sound a lot like Swift functions; they do, and for good reason. Functions in swift are just special cases of closures which are required to have a name and are defined using a slightly different syntax.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/introduction.md:3]: Maybe remove the first bit?

```suggestion
Closures may sound a lot like Swift functions; they do, and for good reason. Functions in swift are just special cases of closures which are required to have a name and are defined using a slightly different syntax.
```

File: languages/swift/exercises/concept/closures/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+As seen above, closures, like regular functions, may omit the `return` keyword if the body of the closure is a single expression.
+
+Most often, closures are used with higher-order functions, either passed in as parameters or returned as results:
+
+```swift
+[11, 75, 3, 99, 53].contains(where: { (x: Int) -> Bool in x > 100 })
+// => false
+
+["apple", "ball", "carrot"].sorted(by: { (s1: String, s2: String) -> Bool in s1.count < s2.count })
+// => ["ball", "apple", "carrot"]
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/introduction.md:42]: This example looks to be quite similar to the above example. Maybe remove this example for brevity?

File: languages/swift/exercises/concept/closures/.docs/introduction.md (hunk 3 above)
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/introduction.md:1]: I like the introduction!

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
+In this exercise you will be using closures with a higher-order function to generate sequences of numbers from a seed number.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:1]: Could you perhaps add a little story/theme to the exercise? It doesn't have to be a big story or anything, but just a little something to give the exercise a bit of story/theme.

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+       generator: (Int) -> (seed: Int, value: Int),
+       stop: (Int) -> Bool
+      ) -> [Int]
+```
+
+`unroll` takes three parameters, a seed value; a `generator` function which, when applied to the seed returns a tuple consisting of the next value for the seed and the next value for the sequence; and a `stop` function which, when applied to the current seed value returns true if the sequence generation is finished. Unroll then returns the generated sequence. For example, to generate the powers of 3 up to 1,000,000, we can write:
+
+```swift
+let powersOf3 = unroll(1,
+                       generator: { (n: Int) -> (seed: Int, next: Int) in (seed: 3 * n, next: n) },
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:16]: Should this be:

```suggestion
                       generator: { (n: Int) -> (seed: Int, next: Int) in (seed: 3 * n, next: n + 1) },
```

?

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                       stop: { (n: Int) -> Bool in n > 1_000_000 })
+// => [1, 3, 9, 27, 81, 243, 729, 2187, 6561, 19683, 59049, 177147, 531441]
+```
+
+You will use unroll to help you generate a few sequences.
+
+## 1. Write helper closures to generate a sequence of digits
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:23]: This task description describes _how_ to implement the task, instead of _what_ to implement (see [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsinstructionsmd) for guidance). This also ties into the story/theme comment I made above.

Note: this comment also applies to the other tasks.

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+One can generate a sequence of digits of a number, _n_ by computing `n % 10` to get the last digit of `n` and then dividing `n` by 10 to remove that last digit. This process can then be repeated until `n` is 0.
+
+For this task, you need to define a closure of type `(Int) -> (seed: Int, next: Int)` That takes a number and returns that number with its last digit removed as `seed:` along with the last digit as `next:`. Assign this closure to a constant named `generateDigit`.
+
+You will also need to generate a closure of type `(Int) -> Bool` That returns `true` if n = 0 and `false` otherwise. Assign this closure to a constant named `stopAtZero`.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:29]:
```suggestion
You will also need to generate a closure of type `(Int) -> Bool` that returns `true` if n = 0 and `false` otherwise. Assign this closure to a constant named `stopAtZero`.
```

File: languages/swift/exercises/concept/closures/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+generateDigit(123)
+// => (seed: 12, next: 3)
+
+stopAtZero(0)
+// => true
+```
+
+## 2. Generate a sequence of digits of an integer
+
+Use these helper closures with unroll to define a closure of type `(Int) -> [Int]` that takes an Int as input and returns the sequence of digits. Assign this closure to a constant named `digitsOf`.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:41]:
```suggestion
Use these helper closures with `unroll` to define a closure of type `(Int) -> [Int]` that takes an `Int` as input and returns the sequence of digits. Assign this closure to a constant named `digitsOf`.
```

File: languages/swift/exercises/concept/closures/.docs/instructions.md (hunk 6 above)
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/instructions.md:1]: There might be a little _too_ much content in these instructions. Is there anything in particular in tasks 5 and 6 that the student has not had to deal with in the previous 4 tasks? If not, I would suggest removing tasks 5 and 6.

File: languages/swift/exercises/concept/closures/.docs/hints.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
+## General
+
+- The syntax for defining a closure is `{ (parameter list) -> ReturnType in bodyOfClosure }`
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/hints.md:3]: Instead of listing the syntax, is there perhaps an external resource you can refer to?

File: languages/swift/exercises/concept/closures/.docs/hints.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+- Closures capture the values in their enclosing scope and can access them without them being passed in.
+- The `stopAtZero` closure can be reused.
+
+## 5. Generate a closure for any base from 2-10
+
+- The closure returned from `generatorForBase(_:)` should function similarly to `generateDigit` for base 10 and `generateBit` for base 2.
+
+## 6. Generate a Collatz sequence from a positive integer
+
+- You can use [`isMultiple(of:)][is-multiple-of] to test if a number is even or odd.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/closures/.docs/hints.md:30]: Maybe rephrase to not directly give away the solution?

**Comment by ErikSchierboom** [comment]
@wneumann Is there anything I can do to help with this PR?

**Comment by wneumann** [comment]
Sorry the last couple of weeks have been really hectic/distracting. My dog tore a ligament in his knee and taking care of him kind of sucked up all my energy for projects outside of work. He got home from surgery yesterday, so I should be able to start working on this again in a couple of days.

**Comment by ErikSchierboom** [comment]
No worries! Life happens. I hope your dog recovers well!

**Comment by wneumann** [comment]
Swapped out the old exercise for a new, more straightforward one. Needs a review from @ErikSchierboom.

**Review by ErikSchierboom — APPROVED** [review]
Great! I like this a lot.
